blackboxopt/optimizers/botorch_base.py

```diff
@@ -166,27 +166,48 @@ def to_numerical(
     return X, Y
 
 
-def init_af_opt_kwargs(af_opt_kwargs: Optional[dict]) -> dict:
-    """Provide default initialization for the acquisition function optimizer
-    configuration. Ensure that all mandatory fields are set to be used by BoTorch.
+def _init_acquisition_function_optimizer(
+    search_space: ps.ParameterSpace,
+    af_opt_kwargs: Optional[dict],
+    torch_dtype: torch.dtype,
+) -> Callable[[AcquisitionFunction], Tuple[torch.Tensor, torch.Tensor]]:
+    """Prepare either BoTorch's `optimize_acqf_discrete` or `optimize_acqf` depending
+    on whether the search space is fully discrete or not and set required defaults if
+    not overridden by `af_opt_kwargs`.
 
     Args:
-        af_opt_kwargs: Acquisition function configuration.
+        search_space: Search space used for optimization.
+        af_opt_kwargs: Acquisition function optimizer configuration.
+        torch_dtype: Torch tensor type.
 
     Returns:
-        Acquisition function optimizer configuration, applicable for BoTorch's
-            optimizer.
+        Acquisition function optimizer that takes an acquisition function and returns a
+        candidate with its associate acquisition function value.
     """
+    kwargs = {} if af_opt_kwargs is None else af_opt_kwargs.copy()
 
-    af_opt_config = {} if af_opt_kwargs is None else af_opt_kwargs
-
-    # number of initial samples during AF optimization
-    af_opt_config.setdefault("raw_samples", 1024)
-
-    # number of restarts during AF optimization
-    af_opt_config.setdefault("num_restarts", 4)
-
-    return af_opt_config
+    is_fully_discrete_space = not any(
+        search_space[n]["parameter"].is_continuous
+        for n in search_space.get_parameter_names()
+    )
+    if is_fully_discrete_space:
```

Comment:
We could also allow for a flag, e.g. `af_opt_method`, with values `lbfgs` and `random` to manually pick this way of optimizing the AF. But this doesn't have to happen in this PR.

Reply:
True, because maybe we have a large integer only search space but with fine granularity and then it might actually make more sense to use lbfgs instead of the random strategy. Let's address it separately as you proposed.